### PR TITLE
Allow for splits involving less than `merge_factor` splits.

### LINKED ISSUE #575: Avoid creating super large segment
Right now the number of segments involved in the last merge to reach maturity follows the same logic as any other merge.

This is lame, because with a  merge factor of 8, we might end up with segments that are almost 8 times larger than the target segment size.

We should have specific logic to allow for the merge of smaller number of segment for that last merge, and ensure that we create segment that are at < 2x merge_max_doc.

### PR DESCRIPTION
**Pull request by fulmicoton**
The point is to bound the resulting split to
`max_merge_docs* 2` instead of `max_merge_docs * merge_factor`.

Closes #575

### CODE REVIEW

File: quickwit-indexing/src/merge_policy.rs [hunk 1]
```diff
@@ -204,19 +208,74 @@ impl StableMultitenantWithTimestampMergePolicy {
...
 
         for (split_ord, split) in splits.iter().enumerate() {
             if split.num_records >= current_level_max_docs {
                 split_levels.push(current_level_start_ord..split_ord);
                 current_level_start_ord = split_ord;
-                current_level_max_docs = 3 * split.num_records;
+                current_level_max_docs = (3 * split.num_records).min(self.max_merge_docs);
+                if split.num_records >= self.max_merge_docs {
+                    return split_levels;
+                }
```
Comment (fmassot) [inline quickwit-indexing/src/merge_policy.rs:221]: this early return sounds very strange and the code will never reach this part, right?

File: quickwit-indexing/src/merge_policy.rs [hunk 2]
```diff
@@ -204,19 +208,74 @@ impl StableMultitenantWithTimestampMergePolicy {
...
+            == MergeCandidateSize::TooSmall
+        {
+            return None;
+        }
+        Some(merge_candidate_start..merge_candidate_end)
+    }
+
+    /// Returns true iff we should stop adding extra split into this
+    /// merge candidate.
+    fn merge_candidate_size(&self, splits: &[SplitMetadata]) -> MergeCandidateSize {
```
Comment (fmassot) [inline quickwit-indexing/src/merge_policy.rs:254]: From what I understand we are checking if we can merge the split candidates. This depends on several conditions: number of splits, number of documents. 
I would rename some things here:


```rust
enum MergeCandidateState {
    NotEnoughSplits,
    Valid,
    TooManySplits, // we can keep too big for both
    TooManyDocuments, // 
}
```

File: quickwit-indexing/src/merge_policy.rs [hunk 3]
```diff
@@ -204,19 +208,74 @@ impl StableMultitenantWithTimestampMergePolicy {
...
+        }
+        if self.merge_candidate_size(&splits[merge_candidate_start..merge_candidate_end])
+            == MergeCandidateSize::TooSmall
+        {
+            return None;
+        }
+        Some(merge_candidate_start..merge_candidate_end)
+    }
+
+    /// Returns true iff we should stop adding extra split into this
```
Comment (fmassot) [inline quickwit-indexing/src/merge_policy.rs:252]:
```suggestion
    /// Returns `MergeCandidateSize` iff we should stop adding extra split into this
```

File: quickwit-indexing/src/merge_policy.rs [hunk 4]
```diff
@@ -204,19 +208,74 @@ impl StableMultitenantWithTimestampMergePolicy {
...
+            return MergeCandidateSize::TooSmall;
+        }
+
+        // There are already enough splits in this merge.
+        if splits.len() >= self.merge_factor_max {
+            return MergeCandidateSize::OneMoreSplitWouldBeTooBig;
+        }
+        let num_docs_in_merge: usize = splits.iter().map(|split| split.num_records).sum();
+
+        // The resulting split will already exceed `max_merge_docs`.
```
Comment (fmassot) [inline quickwit-indexing/src/merge_policy.rs:267]:
```suggestion
        // The resulting split will exceed `max_merge_docs`.
```

**Review by fmassot — APPROVED** [review]
Sounds good, I would rename some things to make it clearer.

File: quickwit-indexing/src/merge_policy.rs (hunk 1 above)
Comment (fulmicoton) [reply quickwit-indexing/src/merge_policy.rs:221]: ah you are absolutely right. It is the leftover of a failed refactorign attempt.

File: quickwit-indexing/src/merge_policy.rs (hunk 2 above)
Comment (fulmicoton) [reply quickwit-indexing/src/merge_policy.rs:254]: From the caller side, the distinction TooManySplits or  TooManyDocuments does not matter though.

I'll add explanations in the rustdoc

Comment (fmassot) [reply quickwit-indexing/src/merge_policy.rs:254]: oh yes, I misunderstood the usage I would have expected a condition like `if splits.len() == self.merge_factor_max` to return `MergeCandidateSize::OneMoreSplitWouldBeTooBig`. 
I guess that the case `splits.len() > self.merge_factor_max` with the current caller implementation.

**Pull request merged**